### PR TITLE
URLWrapper.updateURL gets a more powerful interface, and use it to clean up nested object updates

### PR DESCRIPTION
**Pull request by adamabeshouse**

### CODE REVIEW

File: src/shared/lib/URLWrapper.ts [hunk 1]
```diff
@@ -221,16 +221,25 @@ export default class URLWrapper<
...
+        _updatedParams:
+            | Partial<QueryParamsType>
+            | ((currentQuery: QueryParamsType) => Partial<QueryParamsType>),
         path: string | undefined = undefined,
         clear = false,
         replace = false
     ) {
+        const updatedParams: Partial<QueryParamsType> =
+            typeof _updatedParams === 'function'
+                ? _updatedParams(JSON.parse(JSON.stringify(this.query)))
```
Comment (alisman) [inline src/shared/lib/URLWrapper.ts:233]: why not use mobx's toJS (with recurse everything on)?

File: src/shared/lib/URLWrapper.ts [hunk 2]
```diff
@@ -221,16 +221,25 @@ export default class URLWrapper<
...
+                : _updatedParams;
+
         // get the names of the props that are designated as session props
         const sessionProps = this.getSessionProps();
 
         // overwrite params onto the existing params
-        const mergedParams = _.assign({}, this.query, updatedParams);
+        const mergedParams = clear
```
Comment (alisman) [inline src/shared/lib/URLWrapper.ts:240]: hmm. so nothing was being done with clear before?

**Comment by alisman** [comment]
@adamabeshouse, this looks good.    i think you and i know should go over the relationship between urlWrapper.query and the actual url just to make we understand it.  And also document it.

File: src/shared/lib/URLWrapper.ts (hunk 2 above)
Comment (adamabeshouse) [reply src/shared/lib/URLWrapper.ts:240]: right, if there was a session involved

**Pull request merged**